accord-core/src/main/java/accord/primitives/KeyRanges.java

```diff
@@ -0,0 +1,395 @@
+package accord.primitives;
+
+import accord.api.Key;
+import accord.api.RoutingKey;
+import accord.utils.SortedArrays;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterators;
+
+import java.util.*;
+
+public class KeyRanges implements Iterable<KeyRange>
+{
+    public static final KeyRanges EMPTY = new KeyRanges(new KeyRange[0]);
+
+    // TODO: fix raw parameterized use
+    final KeyRange[] ranges;
+
+    public KeyRanges(KeyRange[] ranges)
+    {
+        Preconditions.checkNotNull(ranges);
+        this.ranges = ranges;
```

Comment:
`Keys.slice` has a requirement that these are sorted, yet I don't see any code that enforces this... the only caller I see that actually builds ranges is `accord.topology.Topology#Topology(long, accord.topology.Shard...)` but that is the user... so should this sort the ranges? And if so how? 

Just pushed tests for `Keys.slice` and hitting the fact that overlapping ranges tricks it, so it starts to filter out even though the ranges include all Ints